File: sdk/go/oidcclients.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

type OIDCClient struct {
	ID                      string   `json:"id"`
	TenantID                string   `json:"tenant_id,omitempty"`
	Name                    string   `json:"name"`
	ClientID                string   `json:"client_id"`
	RedirectURIs            []string `json:"redirect_uris"`
	GrantTypes              []string `json:"grant_types"`
	Scopes                  []string `json:"scopes"`
	TokenEndpointAuthMethod string   `json:"token_endpoint_auth_method,omitempty"`
	CreatedAt               string   `json:"created_at"`
	UpdatedAt               string   `json:"updated_at,omitempty"`
}

type CreateOIDCClientInput struct {
	Name                    string   `json:"name"`
	TenantID                string   `json:"tenant_id,omitempty"`
	RedirectURIs            []string `json:"redirect_uris"`
	GrantTypes              []string `json:"grant_types,omitempty"`
	Scopes                  []string `json:"scopes,omitempty"`
	TokenEndpointAuthMethod string   `json:"token_endpoint_auth_method,omitempty"`
}

type UpdateOIDCClientInput struct {
	Name                    *string  `json:"name,omitempty"`
	RedirectURIs            []string `json:"redirect_uris,omitempty"`
	GrantTypes              []string `json:"grant_types,omitempty"`
	Scopes                  []string `json:"scopes,omitempty"`
	TokenEndpointAuthMethod *string  `json:"token_endpoint_auth_method,omitempty"`
}

type OIDCClientPage struct {
	Data       []OIDCClient
	NextCursor string
}

type OIDCRotateSecretResult struct {
	ClientID     string `json:"client_id"`
	ClientSecret string `json:"client_secret"`
}

type OIDCClients struct{ http *httpClient }

func (r *OIDCClients) Create(ctx context.Context, in CreateOIDCClientInput) (*OIDCClient, error) {
	var out OIDCClient
	err := r.http.do(ctx, http.MethodPost, "/v1/oidc/clients", nil, in, &out, false)
	return &out, err
}

func (r *OIDCClients) Get(ctx context.Context, id string) (*OIDCClient, error) {
	var out OIDCClient
	err := r.http.do(ctx, http.MethodGet, "/v1/oidc/clients/"+url.PathEscape(id), nil, nil, &out, true)
	return &out, err
}

func (r *OIDCClients) Update(ctx context.Context, id string, in UpdateOIDCClientInput) (*OIDCClient, error) {
	var out OIDCClient
	err := r.http.do(ctx, http.MethodPatch, "/v1/oidc/clients/"+url.PathEscape(id), nil, in, &out, false)
	return &out, err
}

func (r *OIDCClients) Delete(ctx context.Context, id string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/oidc/clients/"+url.PathEscape(id), nil, nil, nil, true)
}

func (r *OIDCClients) RotateSecret(ctx context.Context, id string) (*OIDCRotateSecretResult, error) {
	var out OIDCRotateSecretResult
	err := r.http.do(ctx, http.MethodPost, "/v1/oidc/clients/"+url.PathEscape(id)+"/rotate-secret", nil, struct{}{}, &out, false)
	return &out, err
}

func (r *OIDCClients) List(ctx context.Context, params ListParams) (*OIDCClientPage, error) {
	q := url.Values{}
	if params.Tenant != "" {
		q.Set("tenant", params.Tenant)
	}
	if params.Limit > 0 {
		q.Set("limit", strconv.Itoa(params.Limit))
	}
	if params.Cursor != "" {
		q.Set("cursor", params.Cursor)
	}
	var env struct {
		Items      []OIDCClient `json:"items"`
		Data       []OIDCClient `json:"data"`
		NextCursor string       `json:"next_cursor"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/oidc/clients", q, nil, &env, true); err != nil {
		return nil, err
	}
	items := env.Items
	if items == nil {
		items = env.Data
	}
	return &OIDCClientPage{Data: items, NextCursor: env.NextCursor}, nil
}

File: sdk/js/node/src/oidcClients.ts
import type { HttpClient } from "./client.js";
import type { ListParams, Page } from "./users.js";

export interface OidcClient {
  id: string;
  tenant_id?: string;
  name: string;
  client_id: string;
  redirect_uris: string[];
  grant_types: string[];
  scopes: string[];
  token_endpoint_auth_method?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreateOidcClientInput {
  name: string;
  tenant_id?: string;
  redirect_uris: string[];
  grant_types?: string[];
  scopes?: string[];
  token_endpoint_auth_method?: string;
}

export interface UpdateOidcClientInput {
  name?: string;
  redirect_uris?: string[];
  grant_types?: string[];
  scopes?: string[];
  token_endpoint_auth_method?: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
  next_cursor?: string;
}

export class OidcClients {
  constructor(private readonly http: HttpClient) {}

  create(input: CreateOidcClientInput): Promise<OidcClient> {
    return this.http.post<OidcClient>("/v1/oidc/clients", input);
  }

  get(id: string): Promise<OidcClient> {
    return this.http.get<OidcClient>(`/v1/oidc/clients/${encodeURIComponent(id)}`);
  }

  update(id: string, input: UpdateOidcClientInput): Promise<OidcClient> {
    return this.http.patch<OidcClient>(`/v1/oidc/clients/${encodeURIComponent(id)}`, input);
  }

  delete(id: string): Promise<void> {
    return this.http.delete<void>(`/v1/oidc/clients/${encodeURIComponent(id)}`);
  }

  rotateSecret(id: string): Promise<{ client_id: string; client_secret: string }> {
    return this.http.post<{ client_id: string; client_secret: string }>(
      `/v1/oidc/clients/${encodeURIComponent(id)}/rotate-secret`,
      {},
    );
  }

  async list(params: ListParams = {}): Promise<Page<OidcClient>> {
    const res = await this.http.get<ListEnvelope<OidcClient>>("/v1/oidc/clients", {
      query: { tenant: params.tenant, limit: params.limit, cursor: params.cursor },
    });
    return { data: res.items ?? res.data ?? [], nextCursor: res.next_cursor };
  }
}
